pandas/core/dtypes/concat.py

```diff
@@ -308,28 +268,10 @@ def _maybe_unwrap(x):
     ordered = False
     if all(first.is_dtype_equal(other) for other in to_union[1:]):
         # identical categories - fastpath
-        categories = first.categories
-        ordered = first.ordered
-
-        if all(first.categories.equals(other.categories) for other in to_union[1:]):
-            new_codes = np.concatenate([c.codes for c in to_union])
-        else:
-            codes = [first.codes] + [
-                recode_for_categories(other.codes, other.categories, first.categories)
-                for other in to_union[1:]
-            ]
-            new_codes = np.concatenate(codes)
+        return Categorical._concat_same_dtype(
```

Comment:
We could also move the full of `union_categoricals` do the categorical array module?

Reply:
that would be my preference too, but trying to keep the already-broad scope/diff limited